system-helper/flatpak-system-helper.c

```diff
@@ -190,6 +452,346 @@ get_sender_pid (GDBusMethodInvocation *invocation)
+  while (TRUE)
+    {
+      if (!g_file_enumerator_iterate (enumerator, &file_info, NULL, NULL, &error))
+        {
+          g_warning ("Error while file enumerator iterating: %s", error->message);
+          break;
+        }
+
+      if (file_info == NULL)
+        break;
+
+      name = g_file_info_get_name (file_info);
+      if (g_file_info_get_file_type (file_info) == G_FILE_TYPE_DIRECTORY &&
+          g_str_has_prefix (name, prefix))
+        {
+          g_autoptr(GFile) cache_dir = g_file_get_child (repo_tmpfile, name);
+
+          if (check_for_repo_containing_partial_pull (cache_dir, commit, out_location))
+            return TRUE;
+        }
+    }
+
+  return FALSE;
+}
+
+static gboolean
+handle_get_child_repo_for_pull (FlatpakSystemHelper   *object,
+                                GDBusMethodInvocation *invocation,
+                                const gchar           *ref,
+                                const gchar           *commit,
+                                const gchar           *arg_installation)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(FlatpakDir) system = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *bindfs_path = NULL;
+
+  g_debug ("GetChildRepoForPull %s", arg_installation);
+
+  system = dir_get_system (arg_installation, &error);
+  if (system == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  bindfs_path = g_find_program_in_path ("bindfs");
+  if (bindfs_path == NULL)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED,
+                                             "Can't find bindfs in the system.");
+      return TRUE;
+    }
+  else
+    {
+      OngoingPull *new_pull;
+      g_autofree gchar *child_repo_path = NULL;
+      g_autofree gchar *user_dir = NULL;
+      g_autofree gchar *root_dir = NULL;
+      g_autofree gchar *flatpak_dir = NULL;
+      g_autofree gchar *arg_user = NULL;
+      g_autofree gchar *out_location = NULL;
+      g_autofree gchar *repo_tmp = NULL;
+      g_auto(GStrv) parts = NULL;
+      guint uid_pull_count = 0;
+      uid_t uid;
+
+      if (!get_connection_uid (invocation, &uid, &error))
+        {
+          g_dbus_method_invocation_return_gerror (invocation, error);
+          return TRUE;
+        }
+
+      uid_pull_count = get_uid_pull_count (uid);
+      if (uid_pull_count > MAX_PULLS_PER_USER)
+        {
+          g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_LIMITS_EXCEEDED,
+                                                 "Maximum number of pulls per user exceeded for UID: %u", uid);
+          return TRUE;
+        }
+
+      flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+      repo_tmp = g_build_filename (flatpak_dir, "repo", "tmp", NULL);
+
+      parts = flatpak_decompose_ref (ref, NULL);
```

Comment:
Must handle NULL here, because ref is untrusted user input. Actually I think you should use "arg_" prefixes for the user supplied arguments for this reason.